feat(aws-sdk): use span links instead of processing spans

seemk opened the pull request:
## Which problem is this PR solving?

Drop SQS processing spans in favor of span links as per [latest specification](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#batch-receiving).

Fixes https://github.com/open-telemetry/opentelemetry-js-contrib/issues/707

## Short description of the changes

Create a span link for each incoming message that has a propagation context.

seemk commented:
@jj22ee

jj22ee commented on plugins/node/opentelemetry-instrumentation-aws-sdk/README.md:
Can you update this doc with your changes and to remove mention of `processing spans`
[https://github.com/open-telemetry/opentelemetry-js-contrib/blob/instrumentation-aw[…]0/plugins/node/opentelemetry-instrumentation-aws-sdk/doc/sqs.md](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/instrumentation-aws-sdk-v0.47.0/plugins/node/opentelemetry-instrumentation-aws-sdk/doc/sqs.md#receivemessage)

seemk replied:
Updated, removed processing spans part.

jj22ee reviewed (commented):
Hi, thanks for addressing the previous issue(s) in this PR. Noting here that this is should be labelled as a **`Breaking Change`** (I see this is generally okay for minor version upgrades for a `v0.x.y` package).

Reading through the context to understand and cherry-picking some points here:
- ([summarized here](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1477#issuecomment-1836903586)) Current implementation of SQS Receive's process spans has issues related to async and has limitations in JavaScript
- ([discussed here](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/707#issuecomment-1932528774)) Latest OTel spec has removed mention of process spans for `Batch Receiving`, enabling this PR to drop the process spans
  - [older v1.20.0 spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/semantic_conventions/messaging.md#batch-receiving) has indicated that a receive span may have process spans, and each process span has a link to their respective producer spans
  - [newer v1.28.0  spec](https://github.com/open-telemetry/semantic-conventions/blob/v1.28.0/docs/messaging/messaging-spans.md#batch-receiving) not only indicates that the receive span itself should have all the links to the producer contexts, but also removes process spans ([creation of process spans are described as optional here](https://github.com/open-telemetry/semantic-conventions/blob/v1.28.0/docs/messaging/messaging-spans.md#consumer-spans))

Since this PR follows updates in the latest spec and [avoids issues summarized here](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1477#issuecomment-1836903586), I don’t mind this change. I also want to know if @blumamir has additional thoughts in regards to these change to drop process spans?

jj22ee commented on plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/sqs.ts:
```diff
@@ -150,45 +147,30 @@ export class SqsServiceExtension implements ServiceExtension {
-              ),
+        const messages: SQS.Message[] = response?.data?.Messages || [];
+
+        span.setAttribute('messaging.batch.message_count', messages.length);
```
I have seen some discussion that argues to stay with hardcoded attributes until they are stable, so maybe we can keep this as-is.

seemk replied:
I added a `src/semconv.ts` file (https://github.com/open-telemetry/opentelemetry-js/tree/main/semantic-conventions#unstable-semconv) and got rid of the `/incubating` import.

seemk commented on plugins/node/opentelemetry-instrumentation-aws-sdk/test/sqs.test.ts:
```diff
@@ -604,4 +304,76 @@ describe('SQS', () => {
     });
   });
+
+  describe('batch receive', () => {
```
Added tests for links to v3 tests (and the attribute check as well)

seemk commented:
Also removed / switched a few semconvs that were completely removed from the specification (https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#messaging-attributes):
* queue URL -> full URL
* removed SEMATTRS_MESSAGING_DESTINATION_KIND
* renamed messaging system name from `aws.sws` to `aws_sqs`

seemk commented:
@jj22ee

trentm commented:
^^ I merged with main (fixing a couple easy conflicts) so we can get the tests running here again.
I'd love to get this in.
I'll also add a `!` to the title to mark this as a breaking change.

trentm commented on packages/instrumentation-aws-sdk/src/services/sqs.ts:
```diff
@@ -75,8 +70,7 @@ export class SqsServiceExtension implements ServiceExtension {
           spanName = `${queueName} receive`;
-          spanAttributes[SEMATTRS_MESSAGING_OPERATION] =
-            MESSAGINGOPERATIONVALUES_RECEIVE;
+          spanAttributes[ATTR_MESSAGING_OPERATION_TYPE] = 'receive';
```
nit: There *are* enum value exports from the unstable semantic-conventions exports that could be used for 'receive' here (and `aws_sqs` above):

```
8674:export const MESSAGING_OPERATION_TYPE_VALUE_CREATE = "create" as const;
8685:export const MESSAGING_OPERATION_TYPE_VALUE_DELIVER = "deliver" as const;
8694:export const MESSAGING_OPERATION_TYPE_VALUE_PROCESS = "process" as const;
8705:export const MESSAGING_OPERATION_TYPE_VALUE_PUBLISH = "publish" as const;
8714:export const MESSAGING_OPERATION_TYPE_VALUE_RECEIVE = "receive" as const;
8723:export const MESSAGING_OPERATION_TYPE_VALUE_SEND = "send" as const;
8732:export const MESSAGING_OPERATION_TYPE_VALUE_SETTLE = "settle" as const;
```

However, this is very minor and could be done in a follow-up PR.
